File: Frontend/src/backend/estructureService.js
import { appFetch } from '../../backend/appFetch';

const BASE_PATH = '/structure';


export const createProject = async (dto) => {
    return await appFetch('POST', `${BASE_PATH}/projects/createProject`, dto);
};

export const getProject = async (id) => {
    return await appFetch('GET', `${BASE_PATH}/projects/${id}`);
};

export const deleteProject = async (id) => {
    return await appFetch('DELETE', `${BASE_PATH}/projects/deleteProject/${id}`);
};

export const addPlot = async (projectId, dto) => {
    return await appFetch('POST', `${BASE_PATH}/projects/${projectId}/createPlot`, dto);
};

export const modifyPlot = async (plotId, dto) => {
    return await appFetch('PUT', `${BASE_PATH}/plots/modify/${plotId}`, dto);
};

export const deletePlot = async (plotId) => {
    return await appFetch('DELETE', `${BASE_PATH}/plots/delete/${plotId}`);
};

export const findPlotById = async (plotId) => {
    return await appFetch('GET', `${BASE_PATH}/plots/${plotId}`);
};

export const addPremise = async (plotId, dto) => {
    return await appFetch('POST', `${BASE_PATH}/plots/${plotId}/createPremise`, dto);
};

export const modifyPremise = async (premiseId, dto) => {
    return await appFetch('PUT', `${BASE_PATH}/premises/modify/${premiseId}`, dto);
};

export const deletePremise = async (premiseId) => {
    return await appFetch('DELETE', `${BASE_PATH}/premises/delete/${premiseId}`);
};

export const findPremiseById = async (premiseId) => {
    return await appFetch('GET', `${BASE_PATH}/premises/findPremise/${premiseId}`);
};

export const findPremisesByFilterName = async (plotId, name) => {
    return await appFetch('GET', `${BASE_PATH}/plots/${plotId}/premises/search?name=${encodeURIComponent(name)}`);
};

export const addCharacterToNetwork = async (plotId, dto) => {
    return await appFetch('POST', `${BASE_PATH}/plots/${plotId}/network/addCharacter`, dto);
};

export const modifyCharacterRoleInNetwork = async (plotId, characterId, dto) => {
    return await appFetch('PUT', `${BASE_PATH}/plots/${plotId}/network/modifyCharacter/${characterId}`, dto);
};

export const deleteCharacterFromNetwork = async (plotId, characterId) => {
    return await appFetch('DELETE', `${BASE_PATH}/plots/${plotId}/network/characters/deleteCharacter/${characterId}`);
};

export const addRelationshipToNetwork = async (plotId, dto) => {
    return await appFetch('POST', `${BASE_PATH}/plots/${plotId}/network/relationships`, dto);
};

export const modifyRelationship = async (plotId, dto) => {
    return await appFetch('PUT', `${BASE_PATH}/plots/${plotId}/network/relationships`, dto);
};

export const deleteRelationship = async (plotId, characterId1, characterId2) => {
    const params = `?characterId1=${characterId1}&characterId2=${characterId2}`;
    return await appFetch('DELETE', `${BASE_PATH}/plots/${plotId}/network/relationships${params}`);
};

File: Backend/src/main/java/com/tfg/truby_writer/model/services/EstructureService.java
package com.tfg.truby_writer.model.services;


import com.tfg.truby_writer.model.entities.Plot;
import com.tfg.truby_writer.model.entities.Premise;
import com.tfg.truby_writer.model.entities.Project;
import com.tfg.truby_writer.model.services.Block; 
import com.tfg.truby_writer.model.entities.Network;
import com.tfg.truby_writer.model.entities.NetworkNode;
import com.tfg.truby_writer.model.entities.NetworkRelationship;
import com.tfg.truby_writer.model.entities.User;



import com.tfg.truby_writer.model.exceptions.DuplicateInstanceException;
import com.tfg.truby_writer.model.exceptions.InstanceNotFoundException; 
import com.tfg.truby_writer.model.exceptions.ProjectPermissionException;
import com.tfg.truby_writer.model.exceptions.ProjectPermissionException;

import java.util.List;
import com.tfg.truby_writer.model.enums.Enums;



public interface EstructureService {

    //PROJECT

    Project createProject(User user, String name, String description) throws DuplicateInstanceException, InstanceNotFoundException;
    
    Project getProject(Long id) throws InstanceNotFoundException;
        
    void deleteProject(User user,Long id) throws InstanceNotFoundException, ProjectPermissionException;

    List<Project> getAllProjectsByUserId(Long userId) throws InstanceNotFoundException;



    // PLOTS

    Plot addPlot(Long projectId, String name, String dramaticSituation) throws DuplicateInstanceException, InstanceNotFoundException;

    Plot modifyPlot(Long plotId, String name, String dramaticSituation, String structWeaknessNeed, String structDesire, String structAdversary, String structPlan, String structStruggle, String structSelfRevelation, String structNewEquilibrium) throws InstanceNotFoundException;

    void deletePlot(Long plotId) throws InstanceNotFoundException;

    List<Plot> getAllPlotsByProjectId(Long projectId) throws InstanceNotFoundException;
    
    Plot findPlotByName( Long projectId, String name) throws InstanceNotFoundException;

    Plot findPlotById(Long plotId) throws InstanceNotFoundException;

    // PREMISES

    Premise addPremise(Long plotId, String premise) throws DuplicateInstanceException, InstanceNotFoundException;

    Premise modifyPremise(Long premiseId, String premise, String narrativePossibilities, String narrativeChallenges, String problems, String foundingPrinciple, String conflict, String moralDecision) throws InstanceNotFoundException;

    void deletePremise(Long premiseId) throws InstanceNotFoundException;

    Premise findPremiseById(Long premiseId) throws InstanceNotFoundException;

    Block<Premise> findPremisesByFilterName(Long plotId, String name) throws InstanceNotFoundException;

    // NETWORK NODES AND RELATIONSHIPS

    public NetworkNode findCharacterNetworkByCharacterId(Long networkId,Long characterId) throws InstanceNotFoundException;

    public NetworkNode addCharactertoNetwork(Long plotId, Long characterId, Enums.Role role) throws DuplicateInstanceException, InstanceNotFoundException;

    public NetworkNode modifyCharacterRoleInNetwork(Long plotId, Long characterId, Enums.Role role) throws InstanceNotFoundException;
    
    public void deleteCharacterfromNetwork(Long plotId, Long characterId) throws InstanceNotFoundException;

  
    public NetworkRelationship findRelationshipByNodeFromAndNodeTo(Long nodeFromId, Long nodeToId) throws InstanceNotFoundException;

    public NetworkRelationship addRelationshiptoNetwork(Long plotId, Long characterId1, Long characterId2, Enums.RelationshipType relationshipType) 
    throws InstanceNotFoundException, DuplicateInstanceException;

    public void deleteRelationship(Long plotId, Long characterId1, Long characterId2) throws InstanceNotFoundException;

    public NetworkRelationship modifyRelationship(Long plotId, Long characterId1, Long characterId2, Enums.RelationshipType relationshipType) throws InstanceNotFoundException;


}
